ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java

```diff
@@ -295,25 +277,45 @@ public boolean deleteType(final String name) {
   private static final class SourceInfo {
 
     private final DataSource source;
-    private final ReferentialIntegrityTableEntry referentialIntegrity;
+    private final Set<SourceName> linkedSources = new ConcurrentHashSet<>();
+    private final Set<SourceName> referentialSources = new ConcurrentHashSet<>();
 
     private SourceInfo(
         final DataSource source
     ) {
       this.source = Objects.requireNonNull(source, "source");
-      this.referentialIntegrity = new ReferentialIntegrityTableEntry();
     }
 
     private SourceInfo(
         final DataSource source,
-        final ReferentialIntegrityTableEntry referentialIntegrity
+        final Set<SourceName> linkedSources,
+        final Set<SourceName> referentialSources
     ) {
       this.source = Objects.requireNonNull(source, "source");
-      this.referentialIntegrity = referentialIntegrity.copy();
+      this.linkedSources.addAll(
+          Objects.requireNonNull(linkedSources, "linkedSources"));
+      this.referentialSources.addAll(
+          Objects.requireNonNull(referentialSources, "referentialSources"));
     }
 
     public SourceInfo copy() {
-      return new SourceInfo(source, referentialIntegrity);
+      return new SourceInfo(source, linkedSources, referentialSources);
+    }
+
+    public SourceInfo copyWith(final DataSource source) {
+      return new SourceInfo(source, linkedSources, referentialSources);
```

Comment:
we should probably make deep copies of the linked/referential sources

Reply:
The copies are done inside the constructor. I decided to do it there because the Maps should be mutable. Although nobody else is calling the constructor with immutable sets, I am used to doing it that way.